### PR TITLE
GRAMEX-83 ⁃ Jd influxdb

### PR DESCRIPTION
**Pull request by jaidevd**
### InfluxDB setup
Run InfluxDB with Docker:

```bash
$ docker pull influxdb:2.0.8
$ docker run --name influxdb -p 8086:8086 influxdb:2.0.8
```

Then go to [http://localhost:8086](http://localhost:8086) and do the initial setup. Create the following:

1. A username / password pair
1. An org
1. A bucket

Then, got to the "Data" menu on the left pane on the InfluxDB UI:

Screenshot from 2021-09-01 10-09-06

Then, in the "Buckets" tab, create a bucket:

Screenshot from 2021-09-01 10-10-42

Finally, in the "Tokens" tab, copy the token for your username:

image

### Gramex Setup
```yaml
url:
  home:
    pattern: /$YAMLURL/
    handler: FormHandler
    kwargs:
      url: influxdb:http://localhost:8086/
      username: jaidevd
      token: $token  # Token copied from earlier
      org: Gramener
      bucket: stocks
      xsrf_cookies: false
      id: _time  # This is mandatory. InfluxDB will not work for any other value of `id`
```

### Adding Data
Sample:

```json
{
  "symbol": [
    "AAPL",
    "GOOG",
    "MSFT"
  ],
  "open": [
    213.429998,
    626.951088,
    30.620001
  ],
  "close": [
    214.009998,
    626.751061,
    30.950001
  ],
  "low": [
    212.380001,
    624.241073,
    30.59
  ],
  "high": [
    214.499996,
    629.511067,
    31.1
  ],
  "volume": [
    123432400,
    3927000,
    38409100
  ],
  "_time": [
    "2014-09-05 15:48:19.162191",
    "2014-09-05 15:48:19.162191",
    "2014-09-05 15:48:19.162191"
  ]
}
```

Download a bigger file [here](https://gramener0-my.sharepoint.com/:u:/g/personal/jaidev_deshpande_gramener_com/ES1akeukSIRLu23JcuB4u1AB0hm-WErWitApgGpWTNvbXw?e=BPbyY8).

Post data in bulk with with

```bash
POST -H "Content-Type: application/json" --data @stocks.json "http://localhost:9988/?measurement=stocks&tags=symbol"
```

Post individual records as:

```bash
POST "http://localhost:9988/?measurement=stocks&tags=symbol&symbol=AAPL&high=1&low=2&volume=3&open=4&close=5"
```

`_time` is optional. If not specified, will be considered as the server time.

### Querying data
```bash

# Get values of the Apple stock in the last one week:
GET "http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d"

# Get only volume values
GET "http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d&_c=volume"

# Get everything except the volume
GET "http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d&_c=-volume"

# Get Apple and Google stocks on the days when they closed at > 100 points.
GET "http://localhost:9988/?bucket=stocks&symbol=AAPL&symbol=GOOG&close>=100"
```

### Deleting Data
Delete data between two timestamps with:

```bash
DELETE "http://localhost:9988/?_measurement=stocks&time=<start>&time=<stop>"
```

Optionally, delete only records with specific tags. E.g - delete AAPL records  between the 8th August and 29th August:
```bash
DELETE  "http://localhost:9988/?_time=2021-08-08T00:00:00Z&_time=2021-08-29T00:00:00Z&symbol=AAPL"
```

┆Issue is synchronized with this [Jira Bug](https://gramenertech.atlassian.net/browse/GRAMEX-83)

### CODE REVIEW

**Comment by LaxmiShivannavar** [comment]
1. POST works fine for 1st time but not for 2nd time 
2. Tag values are considered as "String" instead of Number 
 
(Happen while Poasting Individual error)

**Comment by jaidevd** [comment]
@LaxmiShivannavar Please take a look now. The earlier curl command for POSTing bulk data was also incorrect. It needed to be modified for Windows. I've done that.

File: gramex/data.py [hunk 1]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+    # InfluxDB Python client doesn't return anything on DELETE, we return a mock number here.
+    return 0
+
+
+def _influxdb_client(url, token, org, **kwargs):
+    from influxdb_client import InfluxDBClient
+
+    url = re.sub(r"^influxdb:", "", url)
+    timeout = kwargs.pop("timeout", 60_000)
+    return InfluxDBClient(url, token, org=org, timeout=timeout, **kwargs)
```
Comment (sanand0) [inline gramex/data.py:1603]: Remove any invalid kwargs -- user config could have any parameters

File: gramex/data.py [hunk 2]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+        db.delete_api().delete(start, stop, predicate, bucket, org)
+    # InfluxDB Python client doesn't return anything on DELETE, we return a mock number here.
+    return 0
+
+
+def _influxdb_client(url, token, org, **kwargs):
+    from influxdb_client import InfluxDBClient
+
+    url = re.sub(r"^influxdb:", "", url)
+    timeout = kwargs.pop("timeout", 60_000)
```
Comment (sanand0) [inline gramex/data.py:1602]: Make this a keyword default argument

File: gramex/data.py [hunk 3]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+        "_fields": [r.get_value() for r in fields.records]
+    }
+
+
+_influxdb_op_map = {"<~": "<=", ">~": ">=", "!": "!=", "": "=="}
+
+
+def _influxdb_offset_limit(controls):
+    offset = controls.pop("_offset", ["-30d"])[0]
+    limit = controls.pop("_limit", [False])[0]
```
Comment (sanand0) [inline gramex/data.py:1525]: Does this offset and limit default give us the LATEST 10,000 records? That's what we need.

File: gramex/data.py [hunk 4]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+            col, agg, op = _filter_col(key, cols)
+            op = _influxdb_op_map.get(op, op)
+            if col in schema["_fields"]:
+                where = " or ".join([f'r._field == "{col}" and r._value {op} {v}' for v in vals])
+            else:
+                where = " or ".join([f'r["{col}"] {op} "{v}"' for v in vals])
+            filters.append(f"|> filter(fn: (r) => {where})")
+        query += "\n".join(filters)
+        if to_drop:
+            print(to_drop)
```
Comment (sanand0) [inline gramex/data.py:1571]: Remove

File: gramex/data.py [hunk 5]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+
+        df = q.query_data_frame(query)
+    return df.drop(["result", "table"], axis=1, errors="ignore")
+
+
+def _delete_influxdb(url, controls, args, org=None, bucket=None, **kwargs):
+    # curl -X DELETE http://localhost:9988/\?_time>=0\&_time<=5
+    with _influxdb_client(url, org=org, **kwargs) as db:
+        schema = _get_influxdb_schema(db, bucket)
+        start, stop = args.pop('_time')
```
Comment (sanand0) [inline gramex/data.py:1585]:
```
        start = args.pop('_time>', ['0'])[0]
        stop = args.pop('_time<', ['99999999999999999'])[0]
```

File: gramex/data.py [hunk 6]
```diff
@@ -1498,14 +1499,178 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+        print(start, stop)
```
Comment (sanand0) [inline gramex/data.py:1586]: Remove

**Comment by LaxmiShivannavar** [comment]
@sanand0 , @jaidevd  : Its been tested for All POST, DELETE and GET commands, And Working as per the expectation

File: gramex/data.py [hunk 7]
```diff
@@ -1595,12 +1594,15 @@ def _delete_influxdb(url, controls, args, org=None, bucket=None, **kwargs):
...
-def _influxdb_client(url, token, org, **kwargs):
+def _influxdb_client(url, token, org, debug=None, timeout=60_000, enable_gzip=False,
+                     default_tags=None):
     from influxdb_client import InfluxDBClient
 
     url = re.sub(r"^influxdb:", "", url)
-    timeout = kwargs.pop("timeout", 60_000)
-    return InfluxDBClient(url, token, org=org, timeout=timeout, **kwargs)
+    return InfluxDBClient(
+        url, token, org=org, debug=None, enable_gzip=False, default_tags=None,
```
Comment (sanand0) [inline gramex/data.py:1603]: debug, enable_gzip, etc should be from kwargs.

**Pull request merged**